File: companion/ZippieCompanionKit/Sources/ZippieCompanionKit/RelayVerdict.swift
import Foundation

/// Whether the relay is running at all, said in terms the Kit can hold.
///
/// `NEVPNStatus` stays in the app: NetworkExtension is an iOS framework and
/// this package builds and tests on a bare toolchain (see Package.swift). The
/// mapping is a switch in the view; everything that DECIDES anything is here,
/// where `swift test` can reach it.
public enum RelayRun: Equatable, Sendable {
    case off
    case starting
    case running
    case stopping
}

/// What the relay screens may honestly say, and nothing more.
///
/// WHY THIS TYPE EXISTS. Until #44 the sentence was computed inside the SwiftUI
/// view, from `cellularReady` - a fact about THIS PHONE'S OWN RADIO - and it
/// read "Connected to the router, waiting for traffic to carry." On 2026-08-07
/// an operator saw that string on a phone that was on no router at all: the
/// router's leg for it was down with `interface: null`, excluded from the bond,
/// never dialled. The screen knew it had STARTED a tunnel; it did not know that
/// anything had answered, and it said the second thing.
///
/// So the rule is evidence, not intent: a sentence about the router is only
/// available once something has ARRIVED from the router. A view that computes
/// its own truth cannot be tested, which is how the wrong string shipped, so
/// the decision lives here next to ProbeVerdict and the copy is pinned by
/// RelayVerdictTests.
public enum RelayVerdict: Equatable, Sendable {
    case off
    case starting
    case stopping
    /// Running, but nothing has reported yet.
    case awaitingFirstReport
    /// A report exists and has gone quiet - the relay was killed and its
    /// counters are a corpse (RelayStatus.stalenessThreshold).
    case notReporting
    case paused(reason: String)
    case noCellular(detail: String?)
    /// Cellular is up and NOTHING has ever arrived from the router.
    case listening
    /// The router is sending, and nothing has left this phone over cellular.
    case notForwarding(detail: String?)
    /// The router sent before and has been silent since.
    case routerQuiet(silentFor: TimeInterval)
    case carrying

    /// How long the router may be silent before the silence is reported.
    ///
    /// PAIRED WITH THE ROUTER'S KEEPALIVE, not picked. `persistent_keepalive`
    /// is 15s (configs/examples/zippie.toml), so a live leg proves itself at
    /// least that often even with no user traffic; in packet datapath the agent
    /// sprays keepalives every tick (`probe_interval_ms`, 500ms). A threshold
    /// under one keepalive interval would report a healthy idle bond as broken.
    /// 25s is the same pairing net.py settled on for keepalive=15, and the cost
    /// of being late here is a sentence, not a blackholed flow.
    public static let routerQuietAfter: TimeInterval = 25

    /// Decide from the evidence.
    ///
    /// ORDER IS THE ARGUMENT. Local faults come first because they are things
    /// this phone genuinely knows and they explain the absence of traffic; only
    /// once cellular is usable does anything get said about the far end.
    public static func evaluate(run: RelayRun,
                                report: RelayStatus?,
                                now: Date = Date(),
                                quietAfter: TimeInterval = RelayVerdict.routerQuietAfter)
        -> RelayVerdict {
        switch run {
        case .off:      return .off
        case .starting: return .starting
        case .stopping: return .stopping
        case .running:  break
        }
        guard let report else { return .awaitingFirstReport }
        if report.isStale(asOf: now) { return .notReporting }

        let stats = report.stats
        if let reason = stats.budgetExhausted { return .paused(reason: reason) }
        guard stats.cellularReady else { return .noCellular(detail: stats.lastError) }

        guard let inbound = stats.lastRouterInboundAt else {
            // No timestamp. Either nothing has ever arrived - the #44 case - or
            // the report came from an older extension binary that did not
            // record one, in which case the forwarded count still PROVES the
            // router sent something and only the WHEN is unknown. Saying "the
            // router has not sent anything" there would be a flat lie, so the
            // count decides and the silence goes unreported until the new
            // binary takes over.
            return stats.upDatagrams > 0 ? .carrying : .listening
        }
        let silence = now.timeIntervalSince(inbound)
        if silence > quietAfter { return .routerQuiet(silentFor: silence) }
        // Inbound proves the router is talking. It does NOT prove this phone
        // got anything out over cellular, and "Carrying" while every upstream
        // send fails is the same fabrication in the other direction.
        return stats.upDatagrams > 0 ? .carrying : .notForwarding(detail: stats.lastError)
    }

    /// The line at the top of the screen.
    public var headline: String {
        switch self {
        case .off:                 return "Off"
        case .starting:            return "Connecting"
        case .stopping:            return "Stopping"
        case .awaitingFirstReport: return "Starting"
        case .notReporting:        return "Not reporting"
        case .paused:              return "Paused"
        case .noCellular:          return "No cellular"
        // NOT "Standing by", which was the old word for this state and reads as
        // "in position, ready to go" - a claim about the pair. This phone is
        // listening and has heard nothing; that is all.
        case .listening:           return "Ready"
        case .notForwarding:       return "Not relaying"
        case .routerQuiet:         return "Router quiet"
        case .carrying:            return "Carrying"
        }
    }

    /// The sentence under it, naming which router when a name is known.
    ///
    /// TAKES A ROUTER NAME BECAUSE OF A SEPARATE PROBLEM FROM THE ONE #44 WAS
    /// FILED FOR. The original bug was claiming a connection the evidence did
    /// not support; that is fixed above by `evaluate` reading a timestamp,
    /// and this parameter changes none of that decision. This is about a
    /// claim that IS supported still being unclear: "the router" reads as the
    /// wifi router this phone is joined to, which on a home network is a
    /// completely different device with nothing to do with zippie ("it says
    /// connected to the router still but... it should be connected to a
    /// zippie router or something", operator feedback, 2026-08-08). Naming
    /// it, or saying "your zippie router" when the name is not known, removes
    /// that ambiguity without asserting anything the evidence does not -
    /// unlike "connected", naming is not a peer-relationship claim, so it
    /// carries no evidentiary bar of its own.
    ///
    /// `router` is expected to be `Settings.routerDisplayName` from the app -
    /// `Settings.routerSSID`, trimmed and nil'd when unset. The Kit stays
    /// blind to where the name comes from, same as everywhere else here.
    /// Defaulted to nil rather than requiring every call site to pass one:
    /// most of this type's tests assert copy that has nothing to do with the
    /// router at all, and forcing them to thread a name through would be
    /// noise.
    public func detail(router: String? = nil) -> String {
        switch self {
        case .off:
            return "This phone is not contributing. Start the relay to lend its "
                 + "cellular to the bond."
        case .starting:
            return "Bringing the tunnel up."
        case .stopping:
            return "Taking the tunnel down."
        case .awaitingFirstReport:
            return "The tunnel is up. Waiting for the relay's first report."
        case .notReporting:
            return "The relay stopped checking in. It was most likely terminated "
                 + "by the system for using too much memory."
        case let .paused(reason):
            return reason
        case let .noCellular(detail):
            return detail ?? "Cellular is not usable right now."
        case .listening:
            // THE SENTENCE #44 IS ABOUT. It used to read "Connected to the
            // router, waiting for traffic to carry." on evidence that consisted
            // entirely of this phone's own cellular being usable.
            return "\(Self.routerSubject(router)) has not sent anything to this phone yet."
        case let .notForwarding(detail):
            guard let detail else {
                return "\(Self.routerSubject(router)) is sending, but nothing has "
                     + "gone out over cellular yet."
            }
            // Dashed, not colon-joined: the errors this carries are themselves
            // "up: no route to host", and a sentence with two colons in it
            // reads as a parsing accident rather than as a sentence.
            return "\(Self.routerSubject(router)) is sending, but nothing has "
                 + "gone out over cellular - \(detail)"
        case let .routerQuiet(silentFor):
            return "\(Self.routerSubject(router)) stopped sending. Last packet "
                 + "\(Self.ago(silentFor)) ago."
        case .carrying:
            return "This phone's cellular is part of the bond."
        }
    }

    /// The subject to open a router sentence with. Named when the operator
    /// has set one; otherwise still says WHAT KIND of router this is about,
    /// rather than leaving "the router" to be read as the wifi router this
    /// phone happens to be joined to. Shared with `ModeDecision.summary(router:)`
    /// in BondMode.swift, which has the identical ambiguity for the same
    /// reason.
    static func routerSubject(_ router: String?) -> String {
        guard let router, !router.isEmpty else { return "Your zippie router" }
        return router
    }

    /// A duration a person reads at a glance. Three-digit seconds are a number,
    /// not a duration. Public because the screen shows the same age next to the
    /// counters, and two spellings of one duration on one screen is the kind of
    /// difference a reader treats as meaningful.
    public static func ago(_ seconds: TimeInterval) -> String {
        let s = Int(seconds.rounded())
        if s < 120 { return "\(s)s" }
        let m = s / 60
        if m < 120 { return "\(m)m" }
        return "\(m / 60)h"
    }

    /// Whether a leg row built from this verdict should read as carrying.
    ///
    /// SHARED RATHER THAN RE-SWITCHED, because two independent switches over
    /// the same verdict are how a widget and a screen end up disagreeing about
    /// one phone's own state. `BondModel`'s single-phone fallback row and the
    /// widget's leg list both call this - see `isDownLeg` below for the other
    /// half of the mapping.
    public var isCarryingLeg: Bool { self == .carrying }

    /// Whether a leg row built from this verdict is a genuine fault rather
    /// than merely idle. `.listening` and `.routerQuiet` are deliberately NOT
    /// here - this phone's radio is fine and the router may simply not have
    /// spoken yet or gone briefly quiet, which `LegRow`'s `.idle` tint (not
    /// `.down`) already treats as unremarkable. Only the states where THIS
    /// PHONE cannot get traffic through count as down.
    public var isDownLeg: Bool {
        switch self {
        case .paused, .noCellular, .notForwarding: return true
        default: return false
        }
    }

    /// Every case, with representative payloads, so a copy rule can be asserted
    /// across all of them at once. Not `CaseIterable` - the associated values
    /// mean there is no single instance per case, and a rule that only checked
    /// the payload-free cases would miss exactly the sentences that vary.
    static let allCasesForCopyReview: [RelayVerdict] = [
        .off, .starting, .stopping, .awaitingFirstReport, .notReporting,
        .paused(reason: "Daily cap of 2 GB reached."),
        .noCellular(detail: nil),
        .noCellular(detail: "cellular unavailable (interface not usable)"),
        .listening,
        .notForwarding(detail: nil),
        .notForwarding(detail: "up: no route to host"),
        .routerQuiet(silentFor: 40),
        .carrying,
    ]
}

File: companion-android/app/src/main/java/app/zippie/companion/RelayVerdict.kt
package app.zippie.companion

/**
 * What the relay screen and notification may honestly say, and nothing more.
 *
 * PORTED FROM ZippieCompanionKit/RelayVerdict.swift (quadseven/zippie#44), same
 * reasoning even though the syntax differs: a sentence about the ROUTER is only
 * available once something has actually ARRIVED from the router.
 *
 * THE DEFECT THIS REPLACES. [RelayStats.summary] used to decide the sentence
 * from counts:
 *
 *     upDatagrams == 0L && downDatagrams == 0L -> "Listening..."
 *     else -> "Carrying for the bond"
 *
 * `upDatagrams`/`downDatagrams` never decrease, so once a single packet had
 * crossed, that rule read "Carrying for the bond" forever - including an hour
 * after the router's leg for this phone went dark. iOS hit exactly this shape
 * of bug from a different local fact (`cellularReady`) and rejected a
 * count-based fix for the same reason: see RelayVerdict.swift and
 * quadseven/zippie#44, "a count alone would not do: upDatagrams never
 * decreases, so one packet an hour ago read as carrying forever."
 *
 * The fix, ported here: [RelayStats.lastRouterInboundAtMs], a TIMESTAMP of the
 * last inbound packet, recorded when it ARRIVES and before any forwarding
 * decision (RelayService.pumpUp) - so a router dialling a phone whose cellular
 * has died still counts as "the router is talking to me".
 *
 * WHAT THIS DOES NOT MODEL. iOS's `RelayRun` (off/starting/running/stopping)
 * comes from `NEVPNStatus`, read in the SwiftUI view and passed into
 * `RelayVerdict.evaluate`. Android has no equivalent signal reaching this
 * layer today - `RelayService`'s foreground-service lifecycle is not threaded
 * through `BondViewModel`/`BondUiState` - and wiring one is out of scope here
 * (those files belong to other in-flight work). So "off" below covers both
 * "never started" and "starting"/"stopping" alike: the only signal available
 * is whether a [RelayReport] exists at all.
 */
sealed class RelayVerdict {
    /** No report exists. Either the relay has never been started, or it was
     *  just stopped and the store was cleared (RelayStatusStore.clear). */
    object Off : RelayVerdict()

    /** A report exists and has gone quiet - the relay was killed (most likely
     *  by the system, for memory) and its counters are a corpse
     *  (RelayReport.STALENESS_MS). */
    object NotReporting : RelayVerdict()

    data class Paused(val reason: String) : RelayVerdict()

    /** The LAN socket never bound. Nothing the router could ever reach, so
     *  nothing below this can honestly be blamed on the router.
     *
     *  Named `cause`, not `detail` - a constructor property cannot reuse the
     *  name of [RelayVerdict.detail] below without clashing with it. */
    data class NotListening(val cause: String?) : RelayVerdict()

    data class NoCellular(val cause: String?) : RelayVerdict()

    /** Cellular is up and the listener is bound, and NOTHING has ever arrived
     *  from the router. */
    object Listening : RelayVerdict()

    /** The router is sending, and nothing has left this phone over cellular. */
    data class NotForwarding(val cause: String?) : RelayVerdict()

    /** The router sent before and has been silent since. */
    data class RouterQuiet(val silentForMs: Long) : RelayVerdict()

    /**
     * This phone is sending, and the ROUTER says nothing has ever arrived from
     * it.
     *
     * THE STATE THAT DID NOT EXIST ON 2026-08-23, and the one the phone was
     * actually in for hours. At the same moment, this screen read:
     *
     *     Carrying
     *     This phone's cellular is part of the bond.
     *     0.2 MB down in 536 datagrams
     *
     * and the router said of that same leg: `never_handshaked: true`,
     * `link_rx_bytes: 0`, `loss_pct: 100`. Both were honest. The phone counted
     * what it SENT; the router counted what ARRIVED; the reply was leaving by
     * the wrong interface (#278).
     *
     * The phone cannot know it is carrying - carrying is a claim about a remote
     * outcome, and the only evidence for it lives at the other end. When the
     * other end is in hand and contradicts the local counters, the contradiction
     * is the more useful thing to draw.
     */
    object RouterSeesNothing : RelayVerdict()

    object Carrying : RelayVerdict()

    /** The line at the top of the screen and the notification title. */
    val headline: String
        get() = when (this) {
            Off -> "Off"
            NotReporting -> "Not reporting"
            is Paused -> "Paused"
            is NotListening -> "Not listening"
            is NoCellular -> "No cellular"
            // NOT "Standing by" - that reads as "in position, ready to go", a
            // claim about the pair. This phone is listening and has heard
            // nothing; that is all it knows (quadseven/zippie#44).
            Listening -> "Ready"
            is NotForwarding -> "Not relaying"
            is RouterQuiet -> "Router quiet"
            // NOT "Carrying, maybe" or "Degraded". The router is not reporting
            // a quality problem - it is reporting that nothing has EVER arrived
            // from this phone, which is a different and more actionable thing.
            RouterSeesNothing -> "Not arriving"
            Carrying -> "Carrying"
        }

    /** The sentence under it, naming which router when a name is known.
     *
     * TAKES A ROUTER NAME FOR A SEPARATE PROBLEM FROM THE ONE #44 WAS FILED
     * FOR. The original bug was claiming a connection the evidence did not
     * support; that stays fixed by [evaluate] reading a timestamp regardless
     * of this parameter. This is about a claim that IS supported still being
     * unclear: "the router" reads as the wifi router this phone is joined
     * to, a different device entirely (#44 operator follow-up, 2026-08-08:
     * "it says connected to the router still but... it should be connected
     * to a zippie router or something"). Ported from
     * ZippieCompanionKit/RelayVerdict.swift's `detail(router:)`, same
     * reasoning: naming is not a peer-relationship claim, so it carries none
     * of the evidentiary bar "connected" did.
     *
     * NOT WIRED TO ANY ANDROID SCREEN YET. Unlike iOS's
     * `Settings.routerSSID`, nothing on this platform reads or stores a
     * router display name - see the note on [RouterProximity] in
     * BondMode.kt for why this app deliberately does not treat SSID as
     * evidence for the mode decision. A display-only name is a different
     * question and could still be read (with ACCESS_FINE_LOCATION) for this
     * parameter alone, but that is unbuilt; wiring it is left for whenever
     * Android grows a source for it, so this parameter exists only to keep
     * this file's copy logic in lockstep with the iOS twin it is ported
     * from - defaulting to null everywhere reproduces the exact text this
     * function always returned. */
    fun detail(router: String? = null): String {
        val who = router?.takeIf { it.isNotEmpty() } ?: "Your zippie router"
        return when (this) {
            Off ->
                "This phone is not contributing. Start the relay to lend its " +
                    "cellular to the bond."
            NotReporting ->
                "The relay stopped checking in. It was most likely killed by " +
                    "the system for using too much memory."
            is Paused -> reason
            is NotListening -> cause ?: "The relay could not bind its listening socket."
            is NoCellular -> cause ?: "Cellular is not usable right now."
            // THE SENTENCE quadseven/zippie#44 IS ABOUT. It used to be derived
            // from cellularReady alone and claimed a connection nothing proved.
            Listening -> "$who has not sent anything to this phone yet."
            is NotForwarding ->
                if (cause == null) {
                    "$who is sending, but nothing has gone out over cellular yet."
                } else {
                    // Dashed, not colon-joined: the errors this carries are
                    // themselves shaped like "up: no route to host", and a
                    // sentence with two colons reads as a parsing accident.
                    "$who is sending, but nothing has gone out over cellular - $cause"
                }
            is RouterQuiet ->
                "$who stopped sending. Last packet ${ago(silentForMs)} ago."
            RouterSeesNothing ->
                "$who has never had anything arrive from this phone. The " +
                    "counters here are real - they count what left, not what " +
                    "landed."
            Carrying -> "This phone's cellular is part of the bond."
        }
    }

    companion object {
        /**
         * How long the router may be silent before the silence is reported.
         *
         * PAIRED WITH THE ROUTER'S KEEPALIVE, not picked. `persistent_keepalive`
         * is 15s (configs/examples/zippie.toml), so a live leg proves itself at
         * least that often even with no user traffic at all; the packet
         * datapath sprays keepalives on top of that every tick
         * (`probe_interval_ms`, 500ms). A threshold under one keepalive
         * interval would report a healthy idle bond as broken. 25s is the same
         * pairing ZippieCompanionKit/RelayVerdict.swift uses for the same
         * keepalive=15 - the cost of being late here is a sentence, not a
         * blackholed flow.
         */
        const val ROUTER_QUIET_AFTER_MS = 25_000L

        /**
         * Decide from the evidence.
         *
         * ORDER IS THE ARGUMENT. Local faults come first because they are
         * things this phone genuinely knows and they explain the absence of
         * traffic; only once cellular is usable and the listener is bound does
         * anything get said about the far end.
         */
        fun evaluate(
            report: RelayReport?,
            nowMs: Long = System.currentTimeMillis(),
            quietAfterMs: Long = ROUTER_QUIET_AFTER_MS,
            /**
             * The router's own verdict on THIS phone's leg: it has never had
             * anything arrive. Read from `never_handshaked` on the leg the
             * console publishes for this phone (BondLegs.Leg.neverAnswered).
             *
             * DEFAULTS FALSE so every caller that has no router view keeps the
             * behaviour it had. Absence of the router's opinion is not evidence
             * against the phone - only a contradiction is, and this flag is
             * only ever true when the router actively said so.
             */
            routerSeesNothing: Boolean = false,
        ): RelayVerdict {
            if (report == null) return Off
            if (report.isStale(nowMs)) return NotReporting

            val stats = report.stats
            stats.budgetExhausted?.let { return Paused(it) }
            if (!stats.cellularReady) return NoCellular(stats.lastError)
            if (!stats.listening) return NotListening(stats.lastError)

            val inbound = stats.lastRouterInboundAtMs
            if (inbound == null) {
                // No timestamp. Either nothing has ever arrived - the #44
                // case - or the report came from an older build that did not
                // record one yet, in which case the forwarded count still
                // PROVES the router sent something and only the WHEN is
                // unknown. Saying "the router has not sent anything" there
                // would be a flat lie, so the count decides until every
                // reporting process has the field.
                if (stats.upDatagrams > 0 && routerSeesNothing) return RouterSeesNothing
                return if (stats.upDatagrams > 0) Carrying else Listening
            }
            val silence = nowMs - inbound
            if (silence > quietAfterMs) return RouterQuiet(silence)
            // Inbound proves the router is talking. It does NOT prove this
            // phone got anything out over cellular, and "Carrying" while every
            // upstream send fails is the same fabrication in the other
            // direction.
            // THE CONTRADICTION OUTRANKS THE LOCAL COUNTERS, and only here.
            // upDatagrams > 0 means this phone forwarded something; the router
            // saying nothing ever arrived means it did not land. Reporting the
            // phone's half alone is what produced "Carrying" over a household
            // outage (#281).
            if (stats.upDatagrams > 0 && routerSeesNothing) return RouterSeesNothing
            return if (stats.upDatagrams > 0) Carrying else NotForwarding(stats.lastError)
        }

        /** A duration a person reads at a glance. Same bucketing as
         *  ZippieCompanionKit/RelayVerdict.swift's `ago(_:)` so the same
         *  silence reads the same on both phones. */
        fun ago(ms: Long): String {
            val s = Math.round(ms / 1000.0)
            if (s < 120) return "${s}s"
            val m = s / 60
            if (m < 120) return "${m}m"
            return "${m / 60}h"
        }

        /** Every case, with representative payloads, so a copy rule can be
         *  asserted across all of them at once. */
        val ALL_CASES_FOR_COPY_REVIEW: List<RelayVerdict> = listOf(
            Off,
            NotReporting,
            Paused("Daily cap of 2 GB reached."),
            NotListening(null),
            NotListening("cannot listen on 51999: address in use"),
            NoCellular(null),
            NoCellular("cellular unavailable (interface not usable)"),
            Listening,
            RouterSeesNothing,
            NotForwarding(null),
            NotForwarding("up: no route to host"),
            RouterQuiet(40_000),
            Carrying,
        )
    }
}
